### PR TITLE
feat (846): added ledger object hashing functionality

### PR DESCRIPTION
**Pull request by TusharPardhe**
## High Level Overview of Change

This PR implements ledger object hashing utilities for XRPL-py, providing functions to compute hashes of various XRPL ledger objects. These utilities are essential for Batch transactions where developers need to know the hash of a ledger entry object before it's created on the ledger.

**Added Functions:**
- `hash_account_root` - Compute hash of AccountRoot ledger entries
- `hash_offer` - Compute hash of Offer ledger entries  
- `hash_escrow` - Compute hash of Escrow ledger entries
- `hash_payment_channel` - Compute hash of Payment Channel ledger entries
- `hash_trustline` (alias: `hash_ripple_state`) - Compute hash of Trustline/RippleState ledger entries
- `hash_signer_list_id` - Compute hash of SignerList ledger entries
- `hash_check` - Compute hash of Check ledger entries
- `hash_ticket` - Compute hash of Ticket ledger entries
- `hash_deposit_preauth` - Compute hash of DepositPreauth ledger entries
- Supporting utility functions: `sha512_half`, `address_to_hex`, `ledger_space_hex`

### Context of Change

As requested in the initial requirements: "Batch can sometimes require you to know the hash of the ledger entry object before processing therefore I would recommend we add the hashing functionality for all ledger objects into this repo so that developers don't need to build their own hashing functions."

This implementation follows the xrpl.js reference implementation to ensure compatibility and correctness. The architecture uses:
- SHA-512 half hashing as per XRPL standards
- Proper ledger space identifiers for each object type
- Correct sequence number formatting (4 bytes = 8 hex characters)
- Address ordering for trustlines to ensure consistent hashing
- Comprehensive error handling for invalid inputs

Alternative approaches considered:
- Using existing cryptographic libraries directly (rejected due to need for XRPL-specific formatting)
- Implementing a single generic hash function (rejected due to different object structures)
- Using xahau-py implementation directly (used as reference but adapted for xrpl-py patterns)

### Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)

### Did you update CHANGELOG.md?

- [x] Yes

## Test Plan

**Test Coverage:**
- 100% line coverage achieved for all hash utility functions
- Tests follow the existing xrpl-py unittest.TestCase pattern
- Comprehensive test suite covering:
  - Basic functionality tests for all hash functions
  - Known test vectors from xrpl.js for compatibility validation
  - Edge cases (zero sequences, large numbers, invalid addresses)
  - Error handling (invalid addresses, invalid ledger spaces)
  - Currency format variations (3-char codes, bytes, hex strings)
  - Address ordering consistency for trustlines
  - Function aliases verification

**Manual Testing:**
1. Run the test suite: `python -m pytest tests/unit/utils/test_hash_utils.py -v`
2. Check coverage: `python -m pytest tests/unit/utils/test_hash_utils.py --cov=xrpl.utils.hash_utils --cov-report=term-missing`
3. Verify hash outputs match expected xrpl.js test vectors for compatibility

**Integration Testing:**
- All functions integrate properly with existing xrpl-py addresscodec functionality
- Hash utilities are properly exported through the main utils module
- No conflicts with existing utility functions

### CODE REVIEW

**Comment by TusharPardhe** [comment]
@dangell7 https://github.com/XRPLF/xrpl-py/issues/846

File: xrpl/utils/hash_utils.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+    Returns:
+        The computed hash of the payment channel in uppercase hexadecimal.
+    """
+    return sha512_half(
+        ledger_space_hex("paychan")
+        + address_to_hex(address)
+        + address_to_hex(dst_address)
+        + _u32_hex(sequence)
+    )
```
[the remark this answers is not in the document]
Comment (ckeshava) [reply xrpl/utils/hash_utils.py:198]: ^^ this appears to be a valid concern. How did you derive the expected sha512 values in the tests? Was it obtained from the cpp implementation (rippled) ?

File: tests/unit/utils/test_hash_utils.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        """Test that hash_offer_id is an alias for hash_offer."""
+        address = "r32UufnaCGL82HubijgJGDmdE5hac7ZvLw"
+        sequence = 137
+        self.assertEqual(hash_offer(address, sequence), hash_offer_id(address, sequence))
+
+    def test_hash_signer_list_id(self):
+        """Test signer list ID hash calculation."""
+        # Test vector from xrpl.js tests
+        address = "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh"
+        expected = "778365D5180F5DF3016817D1F318527AD7410D83F8636CF48C43E8AF72AB49BF"
```
Comment (ckeshava) [inline tests/unit/utils/test_hash_utils.py:117]: These tests have a tight-coupling with rippled. Client libraries need to ensure parity and conformity with the existing rippled implementation.

A better test for these newly proposed methods would be to validate them against rippled client. Please set up integration tests which create the said ledger objects. Use the proposed `hash-<LedgerObject>` methods to retrieve the created ledger objects through their index. Such a test correctly demonstrates the conformity with rippled.

At present time, I cannot evaluate the correctness of the `expected` values in these unit tests.

File: xrpl/utils/hash_utils.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+from xrpl.core import addresscodec
+
+# Constants
+BYTE_LENGTH = 4  # 4 bytes for sequence numbers
+SEQ_HEX_LEN = BYTE_LENGTH * 2  # 8 hex chars
+
+# Ledger space dictionary mapping ledger object types to their namespace identifiers
+LEDGER_SPACES = {
+    "account": "a",
```
Comment (ckeshava) [inline xrpl/utils/hash_utils.py:25]: The [rippled implementation](https://github.com/XRPLF/rippled/blob/17a2606591cf4f45b03b128630895ebef66b42bd/src/libxrpl/protocol/Indexes.cpp#L66C12-L66C27) contains 32 + 3 (deprecated)  = 35 unique values within the `LedgerNameSpace`. Why does the xrpl-py implementation only contain 17 values? Are you planning to add the remaining in later commits?

**Comment by ckeshava** [comment]
@TusharPardhe  can you address the pending PR comments and rebase this work?
